<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<title>Load data</title>

<link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/highlightjs/cdn-release@11.6.0/build/styles/github.min.css">
<script src="https://cdn.jsdelivr.net/combine/gh/highlightjs/cdn-release@11.6.0/build/highlight.min.js,npm/@xiee/utils/js/load-highlight.js" async></script>



<style type="text/css">
body, td {
  font-family: sans-serif;
  background-color: white;
  font-size: 13px;
}
body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 1.5;
  box-sizing: border-box;
}
*, *:before, *:after {
  box-sizing: inherit;
}
tt, code, pre {
  font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}
a:visited { color: #80007f; }
pre, img { max-width: 100%; }
code {
  font-size: 92%;
  border: 1px solid #ccc;
}
code[class] { background-color: #F8F8F8; }
code.language-undefined { background-color: inherit; }
table {
  margin: auto;
  border-top: 1px solid #666;
  border-bottom: 1px solid #666;
}
table thead th { border-bottom: 1px solid #ddd; }
th, td { padding: 5px; }
thead, tfoot, tr:nth-child(even) { background: #eee; }
blockquote {
  color:#666;
  margin:0;
  padding-left: 1em;
  border-left: 0.5em #eee solid;
}
hr { border: 1px #ddd dashed; }
.frontmatter { text-align: center; }

@media print {
  * {
    background: transparent !important;
    color: black !important;
    filter:none !important;
  }
  body {
    font-size: 12pt;
    max-width: 100%;
  }
  a, a:visited { text-decoration: underline; }
  hr {
    visibility: hidden;
    page-break-before: always;
  }
  pre, blockquote {
    padding-right: 1em;
    page-break-inside: avoid;
  }
  tr, img { page-break-inside: avoid; }
  img { max-width: 100% !important; }
  @page :left { margin: 15mm 20mm 15mm 10mm; }
  @page :right { margin: 15mm 10mm 15mm 20mm; }
  p, h2, h3 { orphans: 3; widows: 3; }
  h2, h3 { page-break-after: avoid; }
}
</style>


</head>

<body>
<div class="include-before">

</div>

<div class="frontmatter">
<div class="title"><h1>Load data</h1></div>
<div class="author"><h2></h2></div>
<div class="date"><h3></h3></div>
</div>

<div class="body">
<p>title: “Adapt a classifier: CHETAH example”
shorttitle: “Adapt a classifier”
author: Ghislain FIEVET <a href="mailto:ghislain.fievet@gmail.com">ghislain.fievet@gmail.com</a>
package: adverSCarial
abstract: &gt;
adverSCarial is an R Package designed for generating and analyzing the vulnerability of scRNA-seq
classifiers to adversarial attacks. The package is versatile and provides a format for integrating
any type of classifier. It offers functions for studying and generating two types of attacks,
min change attack and max change attack. The min change attack involves making a small modification
to the input to alter the classification. The max change attack involves making a large modification
to the input without changing its classification.
The package provides a comprehensive solution for evaluating the robustness of scRNA-seq classifiers
against adversarial attacks.
output:
BiocStyle::html_document:
toc: true
toc_depth: 2
vignette: &gt;
%\VignetteIndexEntry{adaptClassifier}
%\VignetteEngine{knitr::knitr}
%\VignetteEncoding{UTF-8}</p>
<h1>Load data</h1>
<pre><code class="language-r">library(adverSCarial)
library(TENxPBMCData)
</code></pre>
<pre><code class="language-r">pbmc &lt;- TENxPBMCData(dataset = &quot;pbmc3k&quot;)
mat_pbmc &lt;- matrixFromSCE(pbmc)

cell_types &lt;- system.file(&quot;extdata&quot;, &quot;pbmc3k_cell_types.tsv&quot;, package=&quot;adverSCarial&quot;)
cell_types &lt;- read.table(cell_types, sep=&quot;\t&quot;)$cell_type
</code></pre>
<h1>Prepare a classifier with <code>CHETAH</code></h1>
<p>Here we demonstrate how to implement a classifier, and take the example of <code>CHETAH</code> a Bioconductor scRNA-seq classifier.</p>
<p>de Kanter JK, Lijnzaad P, Candelli T, Margaritis T, Holstege FCP (2019). “CHETAH: a selective, hierarchical cell type identification method for single-cell RNA sequencing.” Nucleic Acids Research. ISSN 0305-1048, doi: 10.1093/nar/gkz543.</p>
<pre><code class="language-r">library(CHETAH)
library(Seurat)
</code></pre>
<p>First let’s load a <code>train</code> and a <code>test</code> dataset.</p>
<pre><code class="language-r"># Load scRNA-seq datasets
train_3k &lt;- TENxPBMCData(dataset = &quot;pbmc3k&quot;)
test_4k &lt;- TENxPBMCData(dataset = &quot;pbmc4k&quot;)

cell_types_3k &lt;- system.file(&quot;extdata&quot;, &quot;pbmc3k_cell_types.tsv&quot;, package=&quot;adverSCarial&quot;)
cell_types_3k &lt;- read.table(cell_types_3k, sep=&quot;\t&quot;)
colData(train_3k)$celltypes &lt;- cell_types_3k$cell_type
</code></pre>
<p>Then we convert the <code>test</code> dataset into a <code>Seurat</code> object for visualization purpose.</p>
<pre><code class="language-r">counts4seurat &lt;- t(matrixFromSCE(test_4k))
pbmc &lt;- CreateSeuratObject(counts = counts4seurat, project = &quot;pbmc4k&quot;, min.cells = 3, min.features = 200)
# Filter, normalize and find features
pbmc[[&quot;percent.mt&quot;]] &lt;- PercentageFeatureSet(pbmc, pattern = &quot;^MT-&quot;)
pbmc &lt;- subset(pbmc, subset = nFeature_RNA &gt; 200 &amp; nFeature_RNA &lt; 2500 &amp; percent.mt &lt; 5)
pbmc &lt;- NormalizeData(pbmc, normalization.method = &quot;LogNormalize&quot;, scale.factor = 10000)
pbmc &lt;- FindVariableFeatures(pbmc, selection.method = &quot;vst&quot;, nfeatures = 2000)
# Scale and run PCA
pbmc &lt;- ScaleData(pbmc, features = rownames(pbmc))
pbmc &lt;- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
# Cluster and visualize
pbmc &lt;- FindNeighbors(pbmc, dims = 1:10)
pbmc &lt;- FindClusters(pbmc, resolution = 0.5)
</code></pre>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 4217
## Number of edges: 146577
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.8969
## Number of communities: 10
## Elapsed time: 0 seconds
</code></pre>
<pre><code class="language-r">pbmc &lt;- RunUMAP(pbmc, dims = 1:10)
</code></pre>
<p>We annotate cells with <code>CHETAH</code>.</p>
<pre><code class="language-r"># Anotation with CHETAH
reference_3k &lt;&lt;- sceConvertToHGNC(train_3k)
input &lt;- SingleCellExperiment(assays = list(counts = as.matrix(pbmc@assays$RNA@counts)))
input &lt;- CHETAHclassifier(input = input, ref_cells = reference_3k)
input &lt;- Classify(input = input, 0.00001)
pbmc@meta.data[['chetah_classification']] &lt;- input$celltype_CHETAH
Idents(pbmc) &lt;- 'chetah_classification'
</code></pre>
<p>And visualize the results</p>
<pre><code class="language-r"># Known markers for each cell type
markers = c(&quot;IL7R&quot;, &quot;CCR7&quot;, &quot;CD14&quot;, &quot;LYZ&quot;, &quot;S100A4&quot;, &quot;MS4A1&quot;, &quot;CD8A&quot;, &quot;FCGR3A&quot;, &quot;MS4A7&quot;,
              &quot;GNLY&quot;, &quot;NKG7&quot;, &quot;FCER1A&quot;, &quot;CST3&quot;, &quot;PPBP&quot;)
DotPlot(pbmc, features=markers)
</code></pre>
<p>![plot of chunk chetah markers](figure/chetah markers-1.png)</p>
<pre><code class="language-r">DimPlot(pbmc, reduction = &quot;umap&quot;, label = TRUE, pt.size = 0.5)
</code></pre>
<p>![plot of chunk chetah dimplot](figure/chetah dimplot-1.png)</p>
<h2>Adapt the classifier</h2>
<p><code>CHETAH</code> is a classifier that, when given a list of RNA values for a cell, returns a specific cell type from that list. We need to adjust the classifier so that it can assign a cell type to a group of cells.</p>
<p>A classifier function has to be formated as follow to be used with <em>adverSCarial</em>:</p>
<pre><code class="language-R">    classifier = function(expr, clusters, target){
                
                
                c(&quot;cell type&quot;, trust_value)
    }
</code></pre>
<p>The matrix <code>expr</code> contains RNA expression values, the list <code>clusters</code> consists of the cluster IDs for each cell in <code>expr</code>, and <code>target</code> is the ID of the cluster for which we want to have a classification. The function returns a vector with the classification result, and a trust indice.</p>
<p>This is how you can adapt <code>CHETAH</code> for <code>adverSCarial</code>.</p>
<pre><code class="language-r">CHETAHClassifier &lt;- function(expr, clusters, target){
    if (!exists(&quot;reference_3k&quot;)) {
        reference_3k &lt;&lt;- train_3k
    }
    input &lt;- SingleCellExperiment(assays = list(counts = t(expr)))
    input &lt;- CHETAHclassifier(input = input, ref_cells = reference_3k)
    input &lt;- Classify(input = input, 0.01)
    final_predictions = input$celltype_CHETAH[clusters == target]
    ratio &lt;- as.numeric(sort(table(final_predictions), decreasing = TRUE)[1]) /
        sum(as.numeric(sort(table(final_predictions), decreasing = TRUE)))
    predicted_class &lt;- names(sort(table(final_predictions), decreasing = TRUE)[1])
    if ( ratio &lt; 0.3){
        predicted_class &lt;- &quot;NA&quot;
    }
    c(predicted_class, ratio)
}
</code></pre>
<p>You can now test <code>CHETAH</code> classifier with <code>adverSCarial</code> tools.</p>
<p>Let’s run a <code>maxChangeAttack</code>.</p>
<pre><code class="language-r">rna_matrix &lt;- t(as.matrix(pbmc@assays$RNA@counts))
clusters_id &lt;- pbmc@meta.data[['chetah_classification']]
adv_max_change &lt;- advMaxChange(rna_matrix, clusters_id, &quot;DC&quot;, CHETAHClassifier,adv_method=&quot;perc99&quot;, max_split_size = 2000)
</code></pre>
<p>Let’s run this attack and verify if it is successful.</p>
<p>First we modify the <code>rna_matrix</code> matrix on the target cluster, on the genes previously determined.</p>
<pre><code class="language-r">mat_max_adver &lt;- advModifications(rna_matrix, adv_max_change, clusters_id, &quot;DC&quot;)
</code></pre>
<p>Then we verify that classification is still <code>DC</code>.</p>
<pre><code class="language-r">rf_result &lt;- CHETAHClassifier(mat_max_adver, clusters_id, &quot;DC&quot;)
</code></pre>
<pre><code class="language-r">rf_result
</code></pre>
<pre><code>## [1] &quot;DC&quot;   &quot;0.72&quot;
</code></pre>
<pre><code class="language-r">sessionInfo()
</code></pre>
<pre><code>## R version 4.2.0 (2022-04-22)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.4 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=fr_FR.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=fr_FR.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=fr_FR.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=fr_FR.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] SeuratObject_4.1.3          Seurat_4.3.0               
##  [3] CHETAH_1.12.1               ggplot2_3.4.1              
##  [5] TENxPBMCData_1.14.0         HDF5Array_1.24.2           
##  [7] rhdf5_2.40.0                DelayedArray_0.22.0        
##  [9] Matrix_1.5-3                SingleCellExperiment_1.18.1
## [11] SummarizedExperiment_1.26.1 Biobase_2.56.0             
## [13] GenomicRanges_1.48.0        GenomeInfoDb_1.32.4        
## [15] IRanges_2.30.1              S4Vectors_0.34.0           
## [17] BiocGenerics_0.42.0         MatrixGenerics_1.8.1       
## [19] matrixStats_0.63.0          adverSCarial_0.99.0        
## [21] BiocStyle_2.24.0           
## 
## loaded via a namespace (and not attached):
##   [1] AnnotationHub_3.4.0           corrplot_0.92                
##   [3] BiocFileCache_2.4.0           igraph_1.4.1                 
##   [5] plyr_1.8.8                    lazyeval_0.2.2               
##   [7] sp_1.6-0                      splines_4.2.0                
##   [9] listenv_0.9.0                 scattermore_0.8              
##  [11] digest_0.6.31                 htmltools_0.5.4              
##  [13] viridis_0.6.2                 fansi_1.0.4                  
##  [15] magrittr_2.0.3                memoise_2.0.1                
##  [17] tensor_1.5                    cluster_2.1.4                
##  [19] ROCR_1.0-11                   globals_0.16.2               
##  [21] Biostrings_2.64.1             spatstat.sparse_3.0-0        
##  [23] colorspace_2.1-0              blob_1.2.3                   
##  [25] rappdirs_0.3.3                ggrepel_0.9.3                
##  [27] xfun_0.37                     dplyr_1.1.0                  
##  [29] crayon_1.5.2                  RCurl_1.98-1.10              
##  [31] jsonlite_1.8.4                spatstat.data_3.0-0          
##  [33] progressr_0.13.0              zoo_1.8-11                   
##  [35] survival_3.5-3                glue_1.6.2                   
##  [37] polyclip_1.10-4               gtable_0.3.1                 
##  [39] zlibbioc_1.42.0               XVector_0.36.0               
##  [41] leiden_0.4.3                  Rhdf5lib_1.18.2              
##  [43] future.apply_1.10.0           abind_1.4-5                  
##  [45] scales_1.2.1                  pheatmap_1.0.12              
##  [47] DBI_1.1.3                     spatstat.random_3.1-3        
##  [49] miniUI_0.1.1.1                Rcpp_1.0.10                  
##  [51] viridisLite_0.4.1             xtable_1.8-4                 
##  [53] reticulate_1.28               bit_4.0.5                    
##  [55] htmlwidgets_1.6.1             httr_1.4.5                   
##  [57] RColorBrewer_1.1-3            ellipsis_0.3.2               
##  [59] ica_1.0-3                     farver_2.1.1                 
##  [61] pkgconfig_2.0.3               uwot_0.1.14                  
##  [63] deldir_1.0-6                  sass_0.4.5                   
##  [65] dbplyr_2.3.1                  utf8_1.2.3                   
##  [67] labeling_0.4.2                tidyselect_1.2.0             
##  [69] rlang_1.0.6                   reshape2_1.4.4               
##  [71] later_1.3.0                   AnnotationDbi_1.58.0         
##  [73] munsell_0.5.0                 BiocVersion_3.15.2           
##  [75] tools_4.2.0                   cachem_1.0.7                 
##  [77] cli_3.6.0                     generics_0.1.3               
##  [79] RSQLite_2.3.0                 ExperimentHub_2.4.0          
##  [81] ggridges_0.5.4                evaluate_0.20                
##  [83] stringr_1.5.0                 fastmap_1.1.1                
##  [85] goftest_1.2-3                 yaml_2.3.7                   
##  [87] bioDist_1.68.0                knitr_1.42                   
##  [89] bit64_4.0.5                   fitdistrplus_1.1-8           
##  [91] purrr_1.0.1                   RANN_2.6.1                   
##  [93] KEGGREST_1.36.3               dendextend_1.16.0            
##  [95] nlme_3.1-162                  pbapply_1.7-0                
##  [97] future_1.31.0                 mime_0.12                    
##  [99] compiler_4.2.0                plotly_4.10.1                
## [101] filelock_1.0.2                curl_5.0.0                   
## [103] png_0.1-8                     interactiveDisplayBase_1.34.0
## [105] spatstat.utils_3.0-1          tibble_3.1.8                 
## [107] bslib_0.4.2                   stringi_1.7.12               
## [109] highr_0.10                    lattice_0.20-45              
## [111] commonmark_1.8.1              markdown_1.5                 
## [113] vctrs_0.5.2                   pillar_1.8.1                 
## [115] lifecycle_1.0.3               rhdf5filters_1.8.0           
## [117] BiocManager_1.30.20           spatstat.geom_3.0-6          
## [119] lmtest_0.9-40                 jquerylib_0.1.4              
## [121] RcppAnnoy_0.0.20              irlba_2.3.5.1                
## [123] data.table_1.14.8             cowplot_1.1.1                
## [125] bitops_1.0-7                  patchwork_1.1.2              
## [127] httpuv_1.6.9                  R6_2.5.1                     
## [129] bookdown_0.32                 promises_1.2.0.1             
## [131] KernSmooth_2.23-20            gridExtra_2.3                
## [133] parallelly_1.34.0             codetools_0.2-19             
## [135] MASS_7.3-57                   withr_2.5.0                  
## [137] sctransform_0.3.5             GenomeInfoDbData_1.2.8       
## [139] parallel_4.2.0                grid_4.2.0                   
## [141] tidyr_1.3.0                   rmarkdown_2.20               
## [143] Rtsne_0.16                    spatstat.explore_3.0-6       
## [145] shiny_1.7.4
</code></pre>

</div>

<div class="include-after">

</div>

<script src="https://cdn.jsdelivr.net/combine/npm/@xiee/utils/js/center-img.min.js" async></script>
<script>

</script>
</body>

</html>
